### PR TITLE
[Site Isolation] Introduce provisional frames to handle load failures and JS during navigation

### PR DESCRIPTION
**Pull request by achristensen07**
#### 6b90cc53702b153ef3d8652580543a296c0eb7c5
<pre>
[Site Isolation] Introduce provisional frames to handle load failures and JS during navigation
<a href="https://bugs.webkit.org/show_bug.cgi?id=273741">https://bugs.webkit.org/show_bug.cgi?id=273741</a>
<a href="https://rdar.apple.com/127553870">rdar://127553870</a>

Reviewed by Brady Eidson.

Until now, to navigate a RemoteFrame we first transitioned it to a LocalFrame then started a load in it.
This worked fine, but it has two fundamental issues.  First, if JS interacts with the Frame between when
the provisional load starts and the server reply is received and the load commits, the JS will be interacting
with a LocalFrame instead of the old RemoteFrame.  Second, if the provisional load fails, we need the original
RemoteFrame to be in the frame tree afterwards.  With main frame provisional load failures we were just
transitioning them back to a new RemoteFrame and that passed existing tests, but with iframes or any
more involved testing we ran into an issue.

To solve this, I give each WebFrame the ability to have a provisional LocalFrame and I use that for loading
instead, then when the load commits I replace the RemoteFrame in the tree with a LocalFrame.  This model allows
us to fix the two issues.

A WebFrame can now have multiple WebCore::Frames, one possibly provisional.  To make this work, I need
WebLocalFrameLoaderClient to not get its LocalFrame from the WebFrame, but have its own reference to its LocalFrame.
To use a non-nullable type (WeakRef&lt;WebCore::LocalFrame&gt;) I made it so that the FrameLoaderClient construction
happens during the constructor of the LocalFrame rather than making the FrameLoaderClient then passing it in
to the LocalFrame constructor.  I did this using ClientCreator.

This makes the frame construction and destruction less symmetric as well as the transitions from remote to local
and local to remote less symmetric, but it is necessary because the provisional state needs to be handled robustly.

I also did a few simple cleanups of code I was touching anyways, such as moving the WebFrame invalidator to
the shared WebFrameLoaderClient from the local and remote clients to reduce duplicate code, removing the
BrowsingContextGroup::remotePageInProcess function that takes a RegistrableDomain because we already have the
WebProcessProxy and don&apos;t need to look it up with what is hopefully always the right URL, and using
toWebLocalFrameLoaderClient instead of assuming the client is not an EmptyFrameLoaderClient.

I added some test infrastructure to have HTTPServer never send a response, which will be needed for many
upcoming tests in this area.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/ProvisionalFrameCreationParameters.h: Renamed from Source/WebKit/Shared/LocalFrameCreationParameters.h.
* Source/WebKit/Shared/ProvisionalFrameCreationParameters.serialization.in: Renamed from Source/WebKit/Shared/LocalFrameCreationParameters.serialization.in.
* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::takeRemotePageInProcessForProvisionalPage):
* Source/WebKit/UIProcess/BrowsingContextGroup.h:
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::initializeWebPage):
(WebKit::ProvisionalPageProxy::didFailProvisionalLoadForFrame):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::prepareForProvisionalNavigationInProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::continueNavigationInNewProcess):
(WebKit::WebPageProxy::webPageIDInProcess const):
(WebKit::WebPageProxy::webPageIDInProcessForDomain const): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchDidCommitLoad):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::coreLocalFrame const):
(WebKit::WebFrame::createProvisionalFrame):
(WebKit::WebFrame::commitProvisionalFrame):
(WebKit::WebFrame::removeFromTree):
(WebKit::WebFrame::transitionToLocal): Deleted.
* Source/WebKit/WebProcess/WebPage/WebFrame.h:
(WebKit::WebFrame::provisionalFrame):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::createProvisionalFrame):
(WebKit::WebPage::commitProvisionalFrame):
(WebKit::WebPage::loadRequest):
(WebKit::WebPage::transitionFrameToLocal): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::TEST(SiteIsolation, NavigateIframeToProvisionalNavigationFailure)): Deleted.
(TestWebKitAPI::TEST(SiteIsolation, PresentationUpdateAfterCrossSiteNavigation)): Deleted.

Canonical link: <a href="https://commits.webkit.org/278444@main">https://commits.webkit.org/278444@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/f9449e74db5e6369272701599762e14ecc0bd11b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50564 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29860 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2873 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53823 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1254 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52867 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36111 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/904 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41222 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52663 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27506 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43538 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22332 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24905 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/800 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8942 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46890 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/860 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55412 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25662 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/769 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48633 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26923 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43690 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47686 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27787 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7324 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26655 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/loader/NavigationScheduler.cpp [hunk 1]
```diff
@@ -289,12 +289,13 @@ class ScheduledHistoryNavigation : public ScheduledNavigation {
     void fire(Frame& frame) override
     {
         // If the destination HistoryItem is no longer in the back/forward list, then we don't proceed.
-        if (!frame.page()->backForward().containsItem(m_historyItem))
+        RefPtr page { frame.page() };
+        if (!page->backForward().containsItem(m_historyItem))
```
Comment (szewai) [inline Source/WebCore/loader/NavigationScheduler.cpp]: It seems `page()` can return null:
`if (!page || !page->backForward().containsItem(m_historyItem))`

File: Source/WebCore/page/Page.cpp [hunk 2]
```diff
@@ -750,15 +750,17 @@ void Page::setOpenedByDOM()
...
-void Page::goToItem(HistoryItem& item, FrameLoadType type, ShouldTreatAsContinuingLoad shouldTreatAsContinuingLoad)
+void Page::goToItem(LocalFrame* localMainFrame, HistoryItem& item, FrameLoadType type, ShouldTreatAsContinuingLoad shouldTreatAsContinuingLoad)
 {
     // stopAllLoaders may end up running onload handlers, which could cause further history traversals that may lead to the passed in HistoryItem
     // being deref()-ed. Make sure we can still use it with HistoryController::goToItem later.
     Ref protectedItem { item };
 
-    RefPtr localMainFrame = dynamicDowncast<LocalFrame>(m_mainFrame);
-    if (!localMainFrame)
+    if (!localMainFrame) {
```
Comment (szewai) [inline Source/WebCore/page/Page.cpp]: if `localMainFrame` cannot be null, it seems the parameter should be `LocalFrame&` or  `const LocalFrame&`instead

File: Source/WebKit/UIProcess/ProvisionalPageProxy.cpp [hunk 3]
```diff
@@ -217,15 +217,13 @@ void ProvisionalPageProxy::initializeWebPage(RefPtr<API::WebsitePolicies>&& webs
...
             m_webPageID = existingRemotePageProxy->pageID();
             m_mainFrame = existingRemotePageProxy->page()->mainFrame();
             m_messageReceiverRegistration.stopReceivingMessages();
             m_messageReceiverRegistration.transferMessageReceivingFrom(existingRemotePageProxy->messageReceiverRegistration(), *this);
-            LocalFrameCreationParameters localFrameCreationParameters {
-                std::nullopt
-            };
-            protectedProcess->send(Messages::WebPage::TransitionFrameToLocal(localFrameCreationParameters, m_page->mainFrame()->frameID()), m_webPageID);
+            ProvisionalFrameCreationParameters parameters { std::nullopt };
+            send(Messages::WebPage::CreateProvisionalFrame(parameters, m_mainFrame->frameID()));
```
Comment (szewai) [inline Source/WebKit/UIProcess/ProvisionalPageProxy.cpp]: Can it be `send(Messages::WebPage::CreateProvisionalFrame({ }, m_mainFrame->frameID()));`?

File: Source/WebKit/UIProcess/WebFrameProxy.cpp [hunk 4]
```diff
@@ -438,10 +438,10 @@ void WebFrameProxy::prepareForProvisionalNavigationInProcess(WebProcessProxy& pr
     }
 
     if (this->process().processID() != process.processID()) {
-        LocalFrameCreationParameters localFrameCreationParameters {
+        ProvisionalFrameCreationParameters provisionalFrameCreationParameters {
             m_provisionalFrame->layerHostingContextIdentifier()
         };
-        process.send(Messages::WebPage::TransitionFrameToLocal(localFrameCreationParameters, frameID()), page()->webPageIDInProcessForDomain(navigationDomain));
+        process.send(Messages::WebPage::CreateProvisionalFrame(provisionalFrameCreationParameters, frameID()), page()->webPageIDInProcess(process));
```
Comment (szewai) [inline Source/WebKit/UIProcess/WebFrameProxy.cpp]: `process.send(Messages::WebPage::CreateProvisionalFrame({ m_provisionalFrame->layerHostingContextIdentifier() }, frameID()), page()->webPageIDInProcess(process));`

File: Source/WebKit/WebProcess/WebPage/WebFrame.cpp [hunk 5]
```diff
@@ -393,33 +397,51 @@ void WebFrame::transitionToRemote(std::optional<WebCore::LayerHostingContextIden
         ownerElement->scheduleInvalidateStyleAndLayerComposition();
 }
 
-void WebFrame::removeFromTree()
+// FIXME: If we cancel a provisional load before it commits, we need another message to destroy this provisional frame. And a test, obvsly.
```
Comment (szewai) [inline Source/WebKit/WebProcess/WebPage/WebFrame.cpp]: Typo "obvsly"; can we move the comment about adding tests to the test file?

File: Source/WebKit/WebProcess/WebPage/WebFrame.cpp [hunk 6]
```diff
@@ -393,33 +397,51 @@ void WebFrame::transitionToRemote(std::optional<WebCore::LayerHostingContextIden
...
+        setLayerHostingContextIdentifier(*parameters.layerHostingContextIdentifier);
+}
+
+void WebFrame::provisionalLoadFailed()
+{
+    if (RefPtr frame = std::exchange(m_provisionalFrame, nullptr)) {
+        static_cast<WebLocalFrameLoaderClient&>(frame->loader().client()).takeFrameInvalidator().release();
+        // FIXME: Add a test that verifies that provisional frames are not accessible via DOMWindow.frames.
+        // We might need to do some tricky things to not add to the frame tree until committing, but still
+        // have the right main and root frame pointers.
```
Comment (szewai) [inline Source/WebKit/WebProcess/WebPage/WebFrame.cpp]: Can we move the comment about adding test to the test file?

**Review by beidson — APPROVED** [review]
Nothing to add on top of Sihui's comments

**Pull request merged**